repo2docker/buildpacks/r.py

```diff
@@ -186,28 +185,64 @@ def get_packages(self):
-                test_date.isoformat()
+    def get_rspm_snapshot_url(self, snapshot_date, max_days_prior=7):
+        for i in range(max_days_prior):
+            snapshots = requests.post(
+                "https://packagemanager.rstudio.com/__api__/url",
+                # Ask for midnight UTC snapshot
+                json={
+                    "repo": "all",
+                    "snapshot": (snapshot_date - datetime.timedelta(days=i)).strftime(
+                        "%Y-%m-%dT00:00:00Z"
+                    ),
+                },
+            ).json()
+            # Construct a snapshot URL that will give us binary packages for Ubuntu Bionic (18.04)
+            if "upsi" in snapshots:
+                return (
+                    "https://packagemanager.rstudio.com/all/__linux__/bionic/"
+                    + snapshots["upsi"]
+                )
+        raise ValueError(
+            "No snapshot found for {} or {} days prior in packagemanager.rstudio.com".format(
+                snapshot_date.strftime("%Y-%m-%d"), max_days_prior
             )
-            r = requests.head(mran_url)
+        )
+
+    def get_mran_snapshot_url(self, snapshot_date, max_days_prior=7):
+        for i in range(max_days_prior):
+            try_date = snapshot_date - datetime.timedelta(days=i)
+            # Fall back to MRAN if packagemanager.rstudio.com doesn't have it
+            url = "https://mran.microsoft.com/snapshot/{}".format(try_date.isoformat())
+            r = requests.head(url)
             if r.ok:
-                return test_date
-            self.log.warning(
-                "Failed to get MRAN snapshot URL %s: %s %s",
-                mran_url,
-                r.status_code,
-                r.reason,
+                return url
+        raise ValueError(
+            "No snapshot found for {} or {} days prior in mran.microsoft.com".format(
+                snapshot_date.strftime("%Y-%m-%d"), max_days_prior
             )
-        r.raise_for_status()
+        )
+
+    def get_cran_mirror_url(self, snapshot_date):
+        # Date after which we will use rspm + binary packages instead of MRAN + source packages
+        rspm_cutoff_date = datetime.date(2022, 1, 1)
+
+        if snapshot_date >= rspm_cutoff_date or self.r_version >= V("4.1"):
+            return self.get_rspm_snapshot_url(snapshot_date)
+        else:
+            return self.get_mran_snapshot_url(snapshot_date)
+
+    def get_devtools_snapshot_url(self):
+        """
+        Return url of snapshot to use for getting devtools install
+
+        devtools is part of our 'core' base install, so we should have some
+        control over what version we install here.
+        """
+        # IRKernel gets into CRAN on Nov 16 2018 (https://packagemanager.rstudio.com/client/#/repos/1/packages/IRkernel),
+        # so we try snapshot to just after that. We use rspm as we want binary packages all the time
```

Comment:
I'm afraid I don't understand what you mean by "we try snapshot to just after that". Do you mean "anytime after that"?

Reply:
I just cleaned up this function to remove the FIXME and be clearer.